### PR TITLE
kill: resync the container if runtime fails

### PR DESCRIPTION
**Pull request by vrothberg**
If the runtime fails to kill the container there is fair chance that the container has transitionted to another state or been removed already. Take the lock and resync the container to check for that to prevent reading old and potentially outdated state.

[NO NEW TESTS NEEDED]

Signed-off-by: Valentin Rothberg <vrothberg@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
Fix a bug where `podman kill` would mistakenly fail when reading outdated state.
```
@containers/podman-maintainers PTAL

### CODE REVIEW

**Review by Luap99 — CHANGES_REQUESTED** [review]
This will deadlock, at least one caller has already the container lock: https://github.com/containers/podman/blob/47bcd10f61fa3d8da36242370c9aa6720f0f983c/libpod/container_api.go#L218-L226

File: libpod/oci_conmon_common.go [hunk 1]
```diff
@@ -371,7 +371,17 @@ func (r *ConmonOCIRuntime) KillContainer(ctr *Container, signal uint, all bool)
 	}
 	if err := utils.ExecCmdWithStdStreams(os.Stdin, os.Stdout, os.Stderr, env, r.path, args...); err != nil {
 		// Update container state - there's a chance we failed because
-		// the container exited in the meantime.
+		// the container exited in the meantime. Note that we need to
+		// take the lock and resync the container.
+		if !ctr.batched {
+			ctr.lock.Lock()
```
Comment (tyler92) [inline libpod/oci_conmon_common.go:377]: This lock is already locked in https://github.com/containers/podman/blob/bc7695a3b297e89b6933f58369d17be8377c199b/libpod/container_api.go#L220 and this is a deadlock. Am I wrong?

Comment (tyler92) [reply libpod/oci_conmon_common.go:377]: Sorry, dup

**Comment by tyler92** [comment]
> This will deadlock, at least one caller has already the container lock:
> 
> https://github.com/containers/podman/blob/47bcd10f61fa3d8da36242370c9aa6720f0f983c/libpod/container_api.go#L218-L226

Yes, SA https://github.com/containers/podman/issues/15492

**Comment by vrothberg** [comment]
> This will deadlock, at least one caller has already the container lock: 

No caller should have the lock at that point.

**Comment by vrothberg** [comment]
> > This will deadlock, at least one caller has already the container lock:
> 
> No caller should have the lock at that point.

Ah, OK.  That Kill needs to be fixed then. Thanks for highlighting that!

File: libpod/oci_conmon_common.go [hunk 2]
```diff
@@ -370,15 +372,7 @@ func (r *ConmonOCIRuntime) KillContainer(ctr *Container, signal uint, all bool)
 		args = append(args, "kill", ctr.ID(), fmt.Sprintf("%d", signal))
 	}
 	if err := utils.ExecCmdWithStdStreams(os.Stdin, os.Stdout, os.Stderr, env, r.path, args...); err != nil {
-		// Update container state - there's a chance we failed because
-		// the container exited in the meantime.
-		if err2 := r.UpdateContainerStatus(ctr); err2 != nil {
```
Comment (mheon) [inline libpod/oci_conmon_common.go:375]: Can't do this, this is going to break things further up the stack re: Sigproxy. The update needs to remain here.

File: libpod/oci_conmon_common.go [hunk 3]
```diff
@@ -370,15 +372,7 @@ func (r *ConmonOCIRuntime) KillContainer(ctr *Container, signal uint, all bool)
...
-			logrus.Infof("Error updating status for container %s: %v", ctr.ID(), err2)
-		}
-		if ctr.ensureState(define.ContainerStateStopped, define.ContainerStateExited) {
-			return fmt.Errorf("%w: %s", define.ErrCtrStateInvalid, ctr.state.State)
```
Comment (mheon) [inline libpod/oci_conmon_common.go:379]: This is also going to break SigProxy. We need this to remain ErrCtrStateInvalid.

Comment (vrothberg) [reply libpod/oci_conmon_common.go:379]: Can you elaborate a bit more?  At least for kill and stop, the container is not locked, so we should not fiddle with the state.

Maybe we need to make this conditional?

Comment (mheon) [reply libpod/oci_conmon_common.go:379]: I can see only one place where KillContainer is run unlocked. I'll fix that.

We should not be touching KillContainer as such.

Comment (vrothberg) [reply libpod/oci_conmon_common.go:379]: You're right. It's unlocked during stop but not kill.

Comment (mheon) [reply libpod/oci_conmon_common.go:379]: https://github.com/containers/podman/pull/16323 catches the last case where KillContainer is run unlocked

Comment (vrothberg) [reply libpod/oci_conmon_common.go:379]: > I can see only one place where KillContainer is run unlocked. I'll fix that.

So you will fix #16142?

Comment (mheon) [reply libpod/oci_conmon_common.go:379]: Sure

**Comment by vrothberg** [comment]
Closing, @mheon wants to take a shot at this.

**Pull request closed without merging**